### DISCUSSION TITLE
AI-Thinker Audi Kit 2.2 A1s A404

### DISCUSSION THREAD

**Discussion by FrankvanKessel1** [pschatzmann/arduino-audio-tools#1784] [Q&A] [answered]
Hallo allerseits,

Ich bin neu hier und bitte entschuldigt, wenn ich mich als Anfänger hier etwas unbedarft äußere.

Nun zu meinem Problem. Ich versuche schon seit mehreren Wochen einen Code in C++ zu schreiben. Ich benutze/versuche viel von Phil Schatzmann, dem ich an dieser Stelle sehr für seine Arbeit danke. Ich möchte einfach nur eine MP3 von der SD-Karte abspielen. Nicht mehr! Nur einen Ton hören...

Ich bin mir nicht sicher ob mein Board A404 von den Vorlagen unterstützt wird. 

Dazu glaube ich wird der SD Karten Leser nicht unterstützt und ich bekomme nachfolgende Fehlermeldungen nach dem kompilieren.

[W] AudioKit.h : 552 - Mode Button ignored because of conflict: 13

[W] AudioKit.h : 560 - Headphone detection ignored because of conflict: 39

Kann mir jemand helfen, bitte...

Grüße Frank

**Comment by pschatzmann** [comment, marked as answer]
Das sind keine Fehler, sondern nur Warnungen, da einige pins doppelt belegt sind.
Versuche zuest den Sinus Ton Test zum Laufen zu brigen: Das hilft Dir zu bestimmen welchen Treiber Du benutzen musst: 
    
    AudioBoardStream kit(AudioKitEs8388V1); // oder AudioKitEs8388V2 oder AudioKitAC101

Wenn das funktioniert, dann kannst Du dich um die [SD Karte kümmern](https://github.com/pschatzmann/arduino-audiokit/wiki/Problems-with-the-SD-Drive-or-some-conflicting-Buttons).

**Reply by FrankvanKessel1** [reply]
Es muss der AudioKitAC101 sein nur mit diesem kann ich kompilieren

**Reply by pschatzmann** [reply]
Nein, kompilieren sollten alle ohne Probleme!

**Comment by FrankvanKessel1** [comment]
hätten Sie da einen kurzen code für mich? Tue mich total schwer...

**Reply by pschatzmann** [reply]
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-generator-audiokit/streams-generator-audiokit.ino

**Reply by FrankvanKessel1** [reply]
habe folgende Lib geladen 
Arduino-audio-driver, audio-tools und lib helix,

 AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);

ist nicht definiert sagt platform io. Habe ich was vergessen?

**Reply by pschatzmann** [reply]
Remove the version information if you have specifed them in lib_deps.
Delete the .pio folder to force the reload of the latest versions.

**Reply by FrankvanKessel1** [reply]
code läuft aber ohne ton

**Reply by pschatzmann** [reply]
So try the other drivers objects

**Reply by FrankvanKessel1** [reply]
sorry sorry tomorrow läuft mit ac101.

erster Ton übrigens seit ich damit arbeite danke danke

**Reply by FrankvanKessel1** [reply]
bekomme Ton mit AC101. danke danke....

**Comment by FrankvanKessel1** [comment]
jetzt Habe ich in Ihren Code die AC 101 eingetragen und möchte einen Song der Hallo.mp3 heißt auf der SD abspielen. Leider vergebens...

Zuerst zeige ich den Code und dann den Debug. 

Code:
/**
 * @file player-sd-audiokit.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/README.md
 * Make sure that the pins are set to off, on, on, off, off
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioBoardStream.h"
#include "AudioTools/AudioLibs/AudioSourceSD.h" // or AudioSourceIdxSD.h
#include "AudioTools/AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
AudioBoardStream kit(AudioKitAC101);
MP3DecoderHelix decoder;  // or change to MP3DecoderMAD
AudioPlayer player(source, kit, decoder);

void next(bool, int, void*) {
   player.next();
}

void previous(bool, int, void*) {
   player.previous();
}

void startStop(bool, int, void*) {
   player.setActive(!player.isActive());
}

void setup() {
  Serial.begin(115200);
  AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  // sd_active is setting up SPI with the right SD pins by calling 
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
  cfg.sd_active = true;
  kit.begin(cfg);

  // setup additional buttons 
  kit.addDefaultActions();
  kit.addAction(kit.getKey(1), startStop);
  kit.addAction(kit.getKey(4), next);
  kit.addAction(kit.getKey(3), previous);


  // setup player
  player.setVolume(0.7);
  player.begin();

  // select file with setPath() or setIndex()
  player.setPath("/");
  player.setIndex(1); // 2nd file

}

void loop() {
  player.copy();
  kit.processActions();

}

Debug:

[E] AudioSourceSD.h : 60 - SD.begin cs=13 failed
[537858][W][sd_diskio.cpp:174] sdCommand(): no token received
[537958][W][sd_diskio.cpp:174] sdCommand(): no token received
[538058][W][sd_diskio.cpp:174] sdCommand(): no token received
[538157][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
[538157][W][sd_diskio.cpp:516] ff_sd_initialize(): GO_IDLE_STATE failed
[538158][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[538168][W][sd_diskio.cpp:174] sdCommand(): no token received
[538273][W][sd_diskio.cpp:174] sdCommand(): no token received
[538373][W][sd_diskio.cpp:174] sdCommand(): no token received
[538472][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
[E] AudioSourceSD.h : 60 - SD.begin cs=13 failed
[539474][W][sd_diskio.cpp:174] sdCommand(): no token received
[539574][W][sd_diskio.cpp:174] sdCommand(): no token received
[539674][W][sd_diskio.cpp:174] sdCommand(): no token received
[539773][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
[539773][W][sd_diskio.cpp:516] ff_sd_initialize(): GO_IDLE_STATE failed
[539774][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[539784][W][sd_diskio.cpp:174] sdCommand(): no token received
[539889][W][sd_diskio.cpp:174] sdCommand(): no token received
[539989][W][sd_diskio.cpp:174] sdCommand(): no token received
[540088][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00

**Reply by pschatzmann** [reply]
Ich denke die DIP Schalter sind falsch eingestellt: Siehe meine erste Antwort und den Kommentar im Sketch!

Laut Fehlermeldung hast Du zumindest CS GPIO 13 für die SD Karte ausgeschaltet. 

ps SPI.begin is nicht nötig, da cfg.sd_active = true; eingestellt ist!

**Comment by FrankvanKessel1** [comment]
Hallo Herr Schatzmann,

ich habe die DIP eingestellt. alles läuft. Klingt etwas nach Blech der klang könnte besser sein. Ist aber vielleicht der Lautsprecher...

Danke Danke Danke...

**Reply by FrankvanKessel1** [reply]
kann der Klang vielleicht mit "MP3DecoderMAD" verbessert werden?
